Added retry for requests used by EVCs

Alopalao opened the pull request:
Closes #483
Closes #184 

### Summary

Creation/deletion of flows error by Service Unavailable can't be stopped since number of requests can go to infinite.
This PR adds two crucial approaches to try to minimize Service Unavailable error results:

#### Add retry to requests made by EVCs (`flow_manager` and `pathfinder`)
The retry is applied with a random timing of 5 to seconds in three instances. Because Service Unavailable happens when multiple requests are send in a short period of time, the randomness is meant to help with concurrent requests.

#### ~Adding `error_status` to EVCs~

#### Extra measures
- Added extra field to the response for POST request when creating an EVC, now looks like  `{"circuit_id": "08352c061d3447", "deployed": false}`.
- Added extra field to the response for PATH request when updating EVC. Now it includes `{'redeployed': false}`
- Added extra string to the response for PATCH request when redeploying EVC. If an error when deleting old flows an extra message with the error will be added to the response of 409.
~- Mechanic with `evc.old_path` has been changed to `evc.old_paths` which is now a list additionally containing all the old paths that failed to be deleted. This error will not change the state of `error_status` nor disable nor deactivate the EVC. Another event `kytos/mef_eline.cleanup_evcs_old_path` will be sent containing the failed EVCs. The handling of this event is now `dynamic_single`.~
- Introduced two new exceptions ~EVCPathNotDeleted~ and EVCPathNotInstalled. These are raised the the methods that catch FlowModException.
- Removed `requests` in favor of `httpx`

### Local Tests
Stress tested with 100 to 140 EVC concurrent deletions and creations. These add up to ~2000 EVCs per test. Tried with 150 requests but at that point `mef_eline` was failing which did not create any issues just lost requests. `api_concurrency_limit` was not changed.

### End-to-End Tests
```
+ python3 -m pytest tests/ --reruns 2 -r fEr
============================= test session starts ==============================
platform linux -- Python 3.11.2, pytest-8.1.1, pluggy-1.5.0
rootdir: /tests
plugins: rerunfailures-13.0, timeout-2.2.0, anyio-4.3.0
collected 266 items

tests/test_e2e_01_kytos_startup.py ..                                    [  0%]
tests/test_e2e_05_topology.py ..................                         [  7%]
tests/test_e2e_06_topology.py ....                                       [  9%]
tests/test_e2e_10_mef_eline.py ..........ss.....x.....x................  [ 24%]
tests/test_e2e_11_mef_eline.py ......                                    [ 26%]
tests/test_e2e_12_mef_eline.py .....Xx.                                  [ 29%]
tests/test_e2e_13_mef_eline.py ....Xs.s.....Xs.s.XXxX.xxxx..X........... [ 44%]
.                                                                        [ 45%]
tests/test_e2e_14_mef_eline.py x                                         [ 45%]
tests/test_e2e_15_mef_eline.py .....                                     [ 47%]
tests/test_e2e_16_mef_eline.py .                                         [ 47%]
tests/test_e2e_20_flow_manager.py ......................                 [ 56%]
tests/test_e2e_21_flow_manager.py ...                                    [ 57%]
tests/test_e2e_22_flow_manager.py ...............                        [ 62%]
tests/test_e2e_23_flow_manager.py ..............                         [ 68%]
tests/test_e2e_30_of_lldp.py .R...                                       [ 69%]
tests/test_e2e_31_of_lldp.py ...                                         [ 70%]
tests/test_e2e_32_of_lldp.py ...                                         [ 71%]
tests/test_e2e_40_sdntrace.py ................                           [ 77%]
tests/test_e2e_41_kytos_auth.py ........                                 [ 80%]
tests/test_e2e_42_sdntrace.py ..                                         [ 81%]
tests/test_e2e_50_maintenance.py ............................            [ 92%]
tests/test_e2e_60_of_multi_table.py .....                                [ 93%]
tests/test_e2e_70_kytos_stats.py .....R...                               [ 96%]
tests/test_e2e_80_pathfinder.py ss......                                 [100%]
```

viniarck commented on models/evc.py:
```diff
@@ -148,12 +152,13 @@ def __init__(self, controller, **kwargs):
         self.backup_links_cache = set()
         self.affected_by_link_at = get_time("0001-01-01T00:00:00")
-        self.old_path = Path([])
+        self.old_paths = [Path([])]
```
I'm not convinced this is a good idea to be a list of paths. The old path was meant to only hold the soon-to-deleted prior path which is either the `failover_path` (if it's been affected) or the  old `current_path`. 

Let me know if you had any other major reason to justify this, as it is I don't think we should keep appending old_paths. Like, either a path when sending flows can fail to remove or install flows, it can end up in the error state, which is a good idea, but trying to store multiple or repeated paths, can be very confusing. 

If you were to model the `error_status: bool` as `status_reason: set[str]`, then you could also map the exceptions you introduce `EVCPathNotDeleted` and `EVCPathNotInstalled` with their unique string representation like what the `status_reason` value should be like, `deployed_error`, `deletion_error`, meaning if what we have is `deployed_error` then it needs to be redeployed later to be fixed, if it was just a `deletion_error` that means it was from a EVC deletion operation and couldn't get deleted, so eventually it should be deleted by a user (or if one day consistency check can try to repeat this op here too). So, notice that just mapping the action that has failed whether flow installed failed or flow removed failed might not be enough since the context isn't captured, what matters the most is how to get out of the error situation or what to do next. Think about

Does that make sense to you? Or do you see any other way that can make it better?

Alopalao replied:
Yes, I agree `old_paths` can become really heavy and hard to be dealt with. This part can be changed back and the old path can be deleted by event instead of request. Or also just send them through an event after the retries.

The motive for this change was because I caught the `Service Unavailable` when `topology.link_down` was sent and all the installed EVCs were trying to delete `old_path` in addition to re-deployments I trigger concurrently. An error when deleting an old path does not change EVC `error_status` (to be changed to str). So it was just the event `kytos/mef_eline.cleanup_evcs_old_path` being in a loop with EVCs retrying to delete their old paths until `flow_manager` starts to response with 200s.

This part has an added complex since `cleanup_evcs_old_path` now has two triggers and also both modify object lists. Now thinking about it, this process alone is not concurrent so it should be fine with just retries.

viniarck commented on db/models.py:
```diff
@@ -139,6 +139,7 @@ class EVCBaseDoc(DocumentBaseModel):
     metadata: Dict = {}
     active: bool
     enabled: bool
+    error_status: bool = False
```
`error_status: bool` can work, but notice that it's not an optimal choice and make it very difficult to evolve later, plus it's not also aligned with the rest of code base where we've standardized to report errors as a set|list of strs (when the error can get accumulated) like status_reason or just a string. Check out my comment below to see if it makes sense. 

Plus, if data model gets augmented let's not forget DB scripts to backfill, this wouldn't break things but we wan't to have the documents with backfilled values too to avoid confusion. But, this can be handled subsequently before the feature gets completely delivered.

Alopalao replied:
Changed to a dictionary. It an error happens it would populated with `{'current_path': 'install'}` indicating an error happened when installing `current_path`.

Alopalao replied:
Deleted `error_status`, response from error installation and deletion is same as it was before.

viniarck commented on models/evc.py:
```diff
@@ -1261,7 +1335,7 @@ def _send_flow_mods(dpid, flow_mods, command='flows', force=False):
 
         """
-
+        #raise FlowModException(str(1234))
```
Leftover prototyped/exploratory work here. Not good.

Alopalao replied:
Deleted

viniarck commented on models/evc.py:
```diff
@@ -1621,17 +1695,6 @@ def is_using_dynamic_path(self):
             return True
         return False
 
-    def deploy_to(self, path_name=None, path=None):
```
Nice to have this cleaned up. Appreciated.

viniarck commented on models/evc.py:
```diff
@@ -1776,7 +1845,7 @@ def handle_interface_link_up(self, interface: Interface):
         self.activate()
         log.info(
-            f"Activating EVC {self.id}. Interfaces: "
+            f"Activating {self.id}. Interfaces: "
```
Any reasons to remove the EVC str? Minor stuff, but it had an impression that it was still very hand to have it, unless you meant to interpolate the `{self}` dunder repr here?

Alopalao replied:
Yes, I meant to changed it to `{self}` which displays the ID and the name.

viniarck commented on models/path.py:
```diff
@@ -122,6 +125,14 @@ def set_controller(cls, controller=None):
+    @retry(
+        stop=stop_after_attempt(3),
+        wait=wait_combine(wait_fixed(3), wait_random(min=0, max=5)),
+        retry=retry_if_result(lambda result: result == []),
```
This makes sense considered how the `get_paths` return was structured, but notice that it's still lacking to handle request errors, like timeout won't be handled correctly, `log.error` would be logged on each retry (during a retry we have log.warning with tenacity already), so instead of the log.error below in the orignal func, it should reraise a retry-able request exc. Since you're at it, let's go ahead and improve this part too. Something similar that you've implemented on of_lldp https://github.com/kytos-ng/of_lldp/blob/master/main.py#L255-L271

Alopalao replied:
Refactored with new exception `PathFinderException`

viniarck commented on models/evc.py:
```diff
@@ -825,7 +830,17 @@ def should_deploy(self, path=None):
         return False
 
-    def deploy_to_path(self, path=None):  # pylint: disable=too-many-branches
+    def error_hapenned(self):
```
Can this method be named more closely for what it does? By just reading `error_happend()` at call sites you have no idea what will happen. Hindering code readability. Maybe something like `disable_set_error(error: str)` -> None? (assuming the error str discussion)

viniarck replied:
Also, thinking more about it generally, `self.disable()` should really only be used when a user needs to admin disable it (with an update operation) or when a deletion via request happen. I think we should maintain the semantics of `disabled` as it always been. 

All the other retriable failable cases, it should only deactivate, and then set the error accordingly that can still recover it from. Basically the recover for any link_down|link_up is always a redeploy to be fully sure. For a request deletion that only delete from one switch but failed on the other switches should still only deactivate too, and then set the reason with something related that it needs to be deleted again (this last part here can get a bit better if we we're to support in a single request a list of dpid as you suggested before, but considering we're wrapping up the release it'll probably not be doable to fully cover that part but we can plan for it incrementally), either way, using a correct status_reason here too I can see it working.

Final piece of advice: I highly recommend you thinking through the cases again, and organize how we'll use the status_reason values. Happy to keep providing these partial reviews to help the PR to land, but for these cases it's recommended we sort it out first to avoid much back and forth and reimplementations.

Alopalao replied:
Name changed with `deactivate_set_error()`. It no longer disables the EVC. So far, I have no noticed any problem with kytos redeploying for motived like `link_(down|up)` for an EVC with errors.

viniarck commented on models/evc.py:
```diff
@@ -825,7 +830,17 @@ def should_deploy(self, path=None):
+        self.disable()
+        self.error_status = True
+
+    def error_deleted(self):
```
I'll recommend that you try to name it more descriptively for what it does, maybe `set_no_errors` or `clean_errors` or something similar?

Alopalao replied:
Changed with `clean_errors`

viniarck commented on models/evc.py:
```diff
@@ -1244,13 +1309,22 @@ def _install_uni_flows(
+    @retry(
+        stop=stop_after_attempt(3),
+        wait=wait_combine(wait_fixed(3), wait_random(min=2, max=7)),
+        retry=retry_if_exception_type(FlowModException),
```
timeouts aren't still handled

Alopalao replied:
Added `timeout=10`

Alopalao replied:
Explored stress test with default value which is 5 seconds, it should be fine with 10. Changed to 30 nonetheless.

viniarck commented on main.py:
```diff
@@ -949,29 +982,44 @@ def handle_evc_deployed(self, event):
         evc.try_setup_failover_path()
 
-    @listen_to("kytos/mef_eline.cleanup_evcs_old_path")
+    @listen_to("kytos/mef_eline.cleanup_evcs_old_path", pool="dynamic_single")
```
Did you hit any other out of order event issues? 

`dynamic_single` is suitable for many cases, but notice here, this handler can be a bit IO heavy since it can handle many EVCs and it would do a request in a single thread, so the IO throughput would be severely impacted in cases where this handle here gets many events queued. Let's try to understand which cases you've seen that were problematic to see what else we can do here.

Alopalao replied:
This changed was because I had two source for `cleanup_evcs_old_path` to be triggered. I changed it back since `old_paths` is now a single path which flows are sent through an event.

viniarck reviewed (changes requested):
Overall, the retries + set an error state is pretty great. But, there are other things that need to be adapted or changed, check out my comments to see if they make sense. 

Also, changelog hasn't been updated, unit tests failing, happy to provide a partial code review though.

Alopalao commented:
New update.

- Errors in the EVC does not longer disable it. This allows for kytos to redeploy if a topology event happens and affects said EVC.
- Removed `requests` usage and replaced with `httpx`
- Removed unused method from Path class `get_best_path()`
- Changed `old_path` to be a single path again. Its flows are sent through an event now.
- Tests are fixed but some line are still missing to be tested.

viniarck commented on models/evc.py:
```diff
@@ -840,7 +856,15 @@ def deploy_to_path(self, path=None):  # pylint: disable=too-many-branches
+            self.remove_current_flows()
+        except EVCPathNotDeleted as err:
+            log.error(f"Error preparing current_path for {self}: {err}")
+            self.deactivate_set_error('current_path', 'delete')
```
I liked how `error_status` ended up structured with the path and which specific action it failed for, `delete` in this case. My concern is if we should really expose at this level or abstract one more, for instance, users can't do anything about current_path via API other than enabling/disable or redeploying it, so in this case, if this deletion failed here, users will need to `redeploy`. So, would be more valuable for users to have an action of `redeploy` or is that already understood if the users see a `{"error_status": {"current_path": "delete"}}`?, or should we try to also include "redeploy" in the structure too? Meaning capture, which path, which operation failure, and what the user needs to next? 

There's no right or wrong, as its is, and as a developer familiar with the code, I can understand, but I can also imagine a network operator seeing this and thinking that the current_path needs to be deleted again, so will they understand they need to reploy? Think about it. I'll think about it too.

viniarck replied:
On an actual EVC deletion, jsut the specific action isn't too ambiguous though, since there's only path deletions involved, but during an operation that has both delete/install, then it has room for ambiguity.

Alopalao replied:
Change for the value to be an explanation about what happened. E.g. {'current_path': 'Error while deleting old path: In "01:2": Service Unavailable'

viniarck reviewed (commented):
Functionality-wise it's almost there, and nicely done eliminating other tech debt along the way like replacing requests with httpx. There are some minor points to refactor and or fix, check out my new comments. 

Other than that, I think it's worth discussing whether `status_error` should also include the high level action that needs to taken, which would be very valuable for users to fix themselves by calling endpoints and/or more automation with consistency check that we can perform the "recovery" action, think about it, let me know what you think and if you can cover that too, if it's worth it. (feel free not to cover unit test for that part until it's sorted out)

viniarck reviewed (commented):
@Alopalao, I'll recommend that you split the part of setting the `error_status` in a new PR, that's great as it is, but I'd like that one to evolve and be ready to also integrate well with consistency check eventually, and we're in the last days of the release, I don't we'll be able to sort it as thoroughly as we'd like to. That said, let's only handle retries here, and leave the rest of the behavior as it is, so that's already and improved, and then you can tackle and further specify the the rest of the error handling in next issue in a next version. At the end of the day, with the right concurrency limit this shouldn't be much of a concern so it's OK to delay this part in an exchange to make it gets as polished and as aligned with other changes, otherwise we might end up having to refactor again some values which would result in more rework and db scripts, OK? So, map another issue and we'll continue with that in a next release, proceed with finish this PR and the other `2024.1` issues, thanks.

Alopalao commented on models/evc.py:
```diff
@@ -711,17 +721,19 @@ def remove_failover_flows(self, exclude_uni_switches=True,
-        switches.add(self.uni_a.interface.switch)
-        switches.add(self.uni_z.interface.switch)
         if not current_path:
+            if not self.current_path:
```
Removed

Alopalao commented on utils.py:
```diff
@@ -159,3 +161,28 @@ def make_uni_list(list_circuits: list) -> list:
+                }
+            )
+
+            print(f"ID -> {dpid}")
```
Removed

Alopalao commented on utils.py:
```diff
@@ -159,3 +161,28 @@ def make_uni_list(list_circuits: list) -> list:
+    """Send flow mods to be deleted or install to flow_manager
+     through an event"""
+    while flow_dict:
+        offset = settings.BATCH_SIZE or None
```
Refactored

Alopalao commented on models/evc.py:
```diff
@@ -825,7 +831,17 @@ def should_deploy(self, path=None):
+        self.error_status.pop(name_path, None)
+
+    # pylint: disable=too-many-branches
+    def deploy_to_path(self, path=Path([])):
```
Rolled back

Alopalao commented on models/evc.py:
```diff
@@ -890,12 +918,17 @@ def deploy_to_path(self, path=None):  # pylint: disable=too-many-branches
+        if (
+                self.failover_path or not self.current_path
+                or (self.current_path
+                    and self.error_status.get('current_path'))
```
Changed

Alopalao replied:
Changed

Alopalao commented:
- Removed `error_status` as a result I had to remove raises from methods that remove any flows. They do `log.error()` now.
- Change timeout to 30 when sending flows

viniarck commented on main.py:
```diff
@@ -949,19 +944,34 @@ def on_cleanup_evcs_old_path(self, event):
+                )
+                event_contents[evc.id] = content
         if event_contents:
+            send_flow_mods_event(self.controller, total_flows, 'delete')
```
@Alopalao, nice improvement here not to have a failable request and send via events.  You can even make this more reliable during a `deletion` if you were to set and parametrize the force=True value (which the request deletion also uses), that way flow_manager deals handling temporarily switch losing connectivity too. Can you set here and augment `send_flow_mods_event` accordingly?

viniarck reviewed (approved):
@Alopalao, great improvements, and I'm glad how this is evolving gradually. I'll leave it pre-approved. Before merging:

- Augment to parametrize force=True and use it during failover clean up

Finally:

- Map a new issue just so the rest of error handling that you started with `error_status` can be further specified and make official in a future release

viniarck commented:
Let's ship it @Alopalao, feel free to merge.
